### DISCUSSION TITLE
"Install" for use in a QEMU virtual machine

### DISCUSSION THREAD

**Discussion by redleg** [kg4vdk/arcos-linux-modules#19] [It's Not In The Manual...Promise!]
The Title asks most of the question.....

I'm running LMDE6 (Linux Mint Debian Edition) with virtmanager used to manage virtual machine.  This machine is  Lenovo T480 with 64G of ram, so can host multiple VMs easily.

I can create a VM and boot from the ISO image, but there is no obvious way to get a persistent install.....   I saw the discussion on  burning the ISO to an internal SSD, but I need to burn the ISO to a qcow2 formatted virtual disk image to allow me to boot the VM and have persistant storage.

Alternatively, is there a way to point the ISO booted system to a hard disk partition?   Booted from the ISO, I have created created and formatted a 10G (I tried ext4 and vfat) partition, but when rebooted, the system doesn't "pick it up" for storage.

Would it be possible to modify the ISO to look for and use a second storage device, perhaps formatted in a specific way, and/or with a specifically named directory to facilitate identification, for persistent storage?

If so, one could easily run in a VM as I desire, or burn the ISO to one partition on a physical disk, and partition a second partition for storage.

The advantage to this mod would be that the ISO could remain immutable, and storage could be anywhere that was formatted and labeled correctly (an HD partition, another USB drive, perhaps even a network share?)

Thoughts?

AL

**Comment by kg4vdk** [comment]
Hey Al,

Thanks for giving it a whirl, and I applaud your efforts! 

This falls well outside the intended use case for arcOS, and I don't plan to incorporate this functionality. That said, I do lots of testing in KVM environments, and your attempts get pretty close to my method.

When I create the VM, I allocate a 16GB USB (or SATA) disk for the machine, as well as an 8GB "transfer" disk. I then boot from the ISO in the virtual CD drive, format and mount the 8GB disk, and use the live OS to download or `scp` the ISO to the 8GB disk. Once the ISO is on the "transfer" disk, I use the "Disks" utility to "Restore Disk Image" using the ISO as the source, and the 16GB disk as the target. Once completed, I power off the VM, adjust the boot priority as needed and remove the ISO from the virtual CD drive, and I can boot to the "installed" arcOS system.

Heads up...I have had poor experiences in VMs regarding the audio pass-through to the DigiRig. It works well enough to test that it "works." But, the transmitted audio is choppy at best. YMMV.

As such, this method is 100% unsupported. Please let me know how you get along with your attempts!

Mike

**Comment by redleg** [comment]
Hi Mike:

First, thanks for the effort that went into this, and for your reply.

I followed your lead as you outlined above, and am able to boot from that virtual HD.  However, on first boot, and reboots, the desktop status indicators in the lower right of the desktop indicate "no persistent storage", and QRV Modules: Unknown.  I've completed the Station Setup, which also does not persist upon reboot.

I feel like I'm missing something .....


AL

**Comment by redleg** [comment]
OK, an update which seems to work, at least to "install" into a virtual environment.  I'm using the virt-manager frontend to QEMU/KVM on a linux host, but a similar approach may work with other virtualization environments.

Burn the arcOS iso to a USB drive.  I used a 64G Sandisk thumb drive.  The 64G is massive overkill, so you may want something smaller.

Boot your computer with the new arcOS USB drive.  It will convert any unused space to an ARCOS-DATA partition, and prompt you to reboot.  On reboot, initialize the system with your callsign, etc.  Connect to the Internet, and allow the system to install the current modules.  I forced this using the appropriate option under arcOS Tools from the menu.  Reboot several times to verify everything is working as expected.

Make sure your tested arcOS USB drive is plugged in.

Now, create a VM.  With Virt-manager, I used the arcOS iso file as my install media, and created a virtual disk 1GB larger than the USB drive created earlier.  Select the option to Customize the Configuration before installing.  Then, click "Add Hardware",  "USB Host Device", and select your USB drive from the options.

Create the VM, and boot it.

Once the arcOS desktop appears, open a terminal window with ALT-CTRL-T.

Type "lsblk" to list the block devices.  You should see (among other things)  sda, which should be the virtual disk you created as part of the installation, and sdb with three partitions sdb1, sdb2, and sdb3, which represent the USB arcOS drive.  You should be able to verify that you are looking at the correct devices with the sizes.  The virtual disk should be 1GB larger if you created it following the instructions above.

In the terminal window, type "sudo su -" to access a root prompt.

From the root prompt, type the following:   "dd if=/dev/sdb of=/dev/sda" and press return.

This command will copy the contents of the USB drive to the virtual disk.  "if" refers to Input File, "of" to the Output file.  This command will take some time, so be patient and allow it to complete.

Once complete, type "lsblk" again, and you should now see that both sda and sdb have the same structure, with three partitions each, identically sized.

Shut the VM down using the arcOS menu.

With the VM not running, reconfigure it as follows:

Change the Boot Options so that only SATA disk 1 is checked.
Disable the SATA CDROM 1 by erasing the Source Path.
Select the USB Host Device, and click the remove button.

Boot the VM.  If everything works correctly, you should boot into arcOS, and be presented with a station setup window containing the information you entered when setting up the USB drive.  The system should have mounted the ARCOS-DATA partition as expected.

I have NOT tested the applications, or the connections to a radio.  As Mike points out, there may be performance issues (depending on your hardware?).  Mapping the connections between the VM and the radio may be "interesting", and will depend on a lot of factors beyond the scope of just getting this working.


I hope this helps someone......

**Comment by kg4vdk** [comment]
Hey Al,

In my adventures, I have come up with something that may be what you were looking for all along!

I'm attaching a script that converts an ISO image to a QCOW2 disk image of a maximum specified size. In my testing this allows for simulating an arcOS install in a KVM/QEMU environment. Since you're already using KVM, you won't need to install any extra tools. The conversion and resizing is handled by `qemu-img convert` and `qemu-img resize`. The same caveats about device passthrough still apply, however. But this should allow for quicker testing, at least. Just rename from `.txt` to `.sh`, and `chmod +x` the script.

Cheers,
Mike

[iso2qcow.txt](https://github.com/user-attachments/files/17529011/iso2qcow.txt)

Usage: `iso2qcow.sh /path/to/ISO/image.iso`

**Comment by redleg** [comment]
Awesome!

This sounds like it will be a LOT simpler to play with, I'll give it a try.

AL
